File: GEMA_andorid/GEMAINVENTORY/app/src/main/java/com/example/gemainventory/ui/settings/ManageUsersFragment.java
package com.example.gemainventory.ui.settings;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.gemainventory.R;
import com.example.gemainventory.api.RetrofitClient;
import com.example.gemainventory.model.Usuario;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ManageUsersFragment extends Fragment {

    private RecyclerView recyclerView;
    private UserAdapter adapter;
    private List<Usuario> userList = new ArrayList<>();
    
    private String currentUserId;
    private int currentUserRol;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_manage_users, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        // Recuperar sesión del usuario actual
        SharedPreferences prefs = requireContext().getSharedPreferences("GemaPrefs", Context.MODE_PRIVATE);
        currentUserId = prefs.getString("user_id", "");
        currentUserRol = prefs.getInt("user_rol", -1);

        view.findViewById(R.id.btn_back_users).setOnClickListener(v -> requireActivity().onBackPressed());

        recyclerView = view.findViewById(R.id.recycler_users);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

        // Inicializar adaptador con lógica de navegación y eliminación
        adapter = new UserAdapter(
                userList, 
                currentUserId, 
                currentUserRol,
                user -> {
                    // Navegar a edición de rol
                    Bundle args = new Bundle();
                    args.putString("userId", user.getIdUsuario());
                    NavHostFragment.findNavController(this)
                            .navigate(R.id.action_users_to_editRole, args);
                },
                user -> {
                    // Acción de eliminar
                    mostrarConfirmacionEliminar(user);
                }
        );
        recyclerView.setAdapter(adapter);

        FloatingActionButton fabAdd = view.findViewById(R.id.fab_add_user);
        fabAdd.setOnClickListener(v -> {
            NavHostFragment.findNavController(this).navigate(R.id.action_users_to_addUser);
        });

        fetchUsers();
    }

    private void mostrarConfirmacionEliminar(Usuario user) {
        new MaterialAlertDialogBuilder(requireContext())
                .setTitle("Eliminar miembro")
                .setMessage("¿Estás seguro de que quieres eliminar a " + user.getNombre() + " del equipo?")
                .setPositiveButton("Eliminar", (dialog, which) -> {
                    eliminarUsuario(user);
                })
                .setNegativeButton("Cancelar", null)
                .show();
    }

    private void eliminarUsuario(Usuario user) {
        RetrofitClient.INSTANCE.getInstance().eliminarUsuario(user.getIdUsuario()).enqueue(new Callback<Void>() {
            @Override
            public void onResponse(Call<Void> call, Response<Void> response) {
                if (response.isSuccessful()) {
                    Toast.makeText(getContext(), "Usuario eliminado correctamente", Toast.LENGTH_SHORT).show();
                    fetchUsers(); // Recargar lista
                } else {
                    // En vez de un error genérico, intentamos leer el mensaje que envía el servidor
                    String errorMsg = "Error: " + response.code();
                    try {
                        if (response.errorBody() != null) {
                            String body = response.errorBody().string();
                            // Intentar extraer el mensaje del JSON de Spring si existe
                            if (body.contains("\"message\":\"")) {
                                errorMsg = body.split("\"message\":\"")[1].split("\"")[0];
                            } else {
                                errorMsg = body;
                            }
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                    Toast.makeText(getContext(), errorMsg, Toast.LENGTH_LONG).show();
                }
            }

            @Override
            public void onFailure(Call<Void> call, Throwable t) {
                Toast.makeText(getContext(), "Error de red al eliminar", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void fetchUsers() {
        RetrofitClient.INSTANCE.getInstance().obtenerUsuarios("").enqueue(new Callback<List<Usuario>>() {
            @Override
            public void onResponse(Call<List<Usuario>> call, Response<List<Usuario>> response) {
                if (response.isSuccessful() && response.body() != null) {
                    userList.clear();
                    userList.addAll(response.body());
                    adapter.notifyDataSetChanged();
                } else {
                    Toast.makeText(getContext(), "Error al cargar equipo", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<List<Usuario>> call, Throwable t) {
                Toast.makeText(getContext(), "Fallo de red", Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: GEMA_andorid/GEMAINVENTORY/app/src/main/java/com/example/gemainventory/ui/settings/ManageUsersFragment.kt
package com.example.gemainventory.ui.settings

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.compose.runtime.*
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.findNavController
import com.example.gemainventory.R
import com.example.gemainventory.api.RetrofitClient
import com.example.gemainventory.model.Usuario
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class ManageUsersFragment : Fragment() {

    private var currentUserId: String = ""
    private var currentUserRol: Int = -1

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val prefs = requireContext().getSharedPreferences("GemaPrefs", Context.MODE_PRIVATE)
        currentUserId = prefs.getString("user_id", "") ?: ""
        currentUserRol = prefs.getInt("user_rol", -1)
        
        val isDarkMode = prefs.getBoolean("DarkMode", false)

        return ComposeView(requireContext()).apply {
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
            setContent {
                var userList by remember { mutableStateOf<List<Usuario>>(emptyList()) }
                var isLoading by remember { mutableStateOf(true) }

                LaunchedEffect(Unit) {
                    fetchUsers { users ->
                        userList = users
                        isLoading = false
                    }
                }

                ManageTeamScreen(
                    darkTheme = isDarkMode,
                    userList = userList,
                    currentUserId = currentUserId,
                    currentUserIdRol = currentUserRol,
                    isLoading = isLoading,
                    onBackClick = {
                        findNavController().popBackStack()
                    },
                    onAddUserClick = {
                        findNavController().navigate(R.id.action_users_to_addUser)
                    },
                    onEditRoleClick = { user ->
                        val args = Bundle().apply {
                            putString("userId", user.idUsuario)
                        }
                        findNavController().navigate(R.id.action_users_to_editRole, args)
                    },
                    onDeleteClick = { user ->
                        mostrarConfirmacionEliminar(user) {
                            isLoading = true
                            fetchUsers { users ->
                                userList = users
                                isLoading = false
                            }
                        }
                    }
                )
            }
        }
    }

    private fun fetchUsers(onResult: (List<Usuario>) -> Unit) {
        RetrofitClient.instance.obtenerUsuarios("").enqueue(object : Callback<List<Usuario>> {
            override fun onResponse(call: Call<List<Usuario>>, response: Response<List<Usuario>>) {
                val users = response.body()
                if (response.isSuccessful && users != null) {
                    val updatedUsers = users.toMutableList()
                    val prefs = requireContext().getSharedPreferences("GemaPrefs", Context.MODE_PRIVATE)
                    val currentUserEmail = prefs.getString("user_correo", "") ?: ""
                    val currentUserId = prefs.getString("user_id", "") ?: ""

                    // Primero, intentamos llenar lo que ya tenemos (ej. usuario actual)
                    users.forEach { user ->
                        if (user.idUsuario == currentUserId && user.correo.isNullOrBlank() && user.email.isNullOrBlank()) {
                            user.correo = currentUserEmail
                        }
                    }

                    // Ahora identificamos quiénes REALMENTE necesitan una llamada al servidor
                    val usersNeedingFetch = users.filter { 
                        it.correo.isNullOrBlank() && it.email.isNullOrBlank() && !it.idUsuario.isNullOrBlank()
                    }
                    
                    if (usersNeedingFetch.isEmpty()) {
                        onResult(updatedUsers)
                        return
                    }

                    var completedCount = 0
                    users.forEachIndexed { index, user ->
                        if (user.correo.isNullOrBlank() && user.email.isNullOrBlank() && !user.idUsuario.isNullOrBlank()) {
                            RetrofitClient.instance.getUsuarioById(user.idUsuario!!).enqueue(object : Callback<Usuario> {
                                override fun onResponse(call: Call<Usuario>, resp: Response<Usuario>) {
                                    val detailedUser = resp.body()
                                    if (resp.isSuccessful && detailedUser != null) {
                                        // MERGE: Solo actualizamos si el detalle trae información útil
                                        val mergedUser = user.copy(
                                            correo = detailedUser.correo ?: detailedUser.email ?: user.correo,
                                            email = detailedUser.email ?: detailedUser.correo ?: user.email,
                                            direccion = detailedUser.direccion ?: user.direccion,
                                            telefono = detailedUser.telefono ?: user.telefono,
                                            idRol = detailedUser.idRol ?: user.idRol,
                                            imagenUrl = detailedUser.imagenUrl ?: user.imagenUrl
                                        )
                                        updatedUsers[index] = mergedUser
                                    }
                                    completedCount++
                                    if (completedCount == usersNeedingFetch.size) onResult(updatedUsers)
                                }

                                override fun onFailure(call: Call<Usuario>, t: Throwable) {
                                    completedCount++
                                    android.util.Log.e("ManageUsers", "Error al obtener detalles de ${user.idUsuario}: ${t.message}")
                                    if (completedCount == usersNeedingFetch.size) onResult(updatedUsers)
                                }
                            })
                        }
                    }
                } else {
                    onResult(emptyList())
                    Toast.makeText(context, "Error al cargar equipo", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<List<Usuario>>, t: Throwable) {
                onResult(emptyList())
                Toast.makeText(context, "Fallo de red", Toast.LENGTH_SHORT).show()
            }
        })
    }

    private fun mostrarConfirmacionEliminar(user: Usuario, onDeleted: () -> Unit) {
        MaterialAlertDialogBuilder(requireContext())
            .setTitle("Eliminar miembro")
            .setMessage("¿Estás seguro de que quieres eliminar a ${user.nombre} del equipo?")
            .setPositiveButton("Eliminar") { _, _ ->
                eliminarUsuario(user, onDeleted)
            }
            .setNegativeButton("Cancelar", null)
            .show()
    }

    private fun eliminarUsuario(user: Usuario, onDeleted: () -> Unit) {
        val id = user.idUsuario ?: return
        RetrofitClient.instance.eliminarUsuario(id).enqueue(object : Callback<Void> {
            override fun onResponse(call: Call<Void>, response: Response<Void>) {
                if (response.isSuccessful) {
                    Toast.makeText(context, "Usuario eliminado correctamente", Toast.LENGTH_SHORT).show()
                    onDeleted()
                } else {
                    var errorMsg = "Error: ${response.code()}"
                    try {
                        val body = response.errorBody()?.string()
                        if (body != null) {
                            if (body.contains("\"message\":\"")) {
                                errorMsg = body.split("\"message\":\"")[1].split("\"")[0]
                            } else {
                                errorMsg = body
                            }
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                    Toast.makeText(context, errorMsg, Toast.LENGTH_LONG).show()
                }
            }

            override fun onFailure(call: Call<Void>, t: Throwable) {
                Toast.makeText(context, "Error de red al eliminar", Toast.LENGTH_SHORT).show()
            }
        })
    }
}
